### ISSUE TITLE
Add official Linux ARM/Raspberry Pi editor and export template binaries

### ISSUE THREAD

**Issue by Schweini07** [godotengine/godot-proposals#988]
***Bugsquad edit:** You can use https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi until official builds are provided.*

___

The re-open of https://github.com/godotengine/godot/issues/2671.

**NOTE**: This issue is more about Godot 3.2, as Godot 3.0.6 and 3.1 are possible to get on the pi, with the costs of a few bugs, for more information read below.

If you have anything to add, please tell me.

**Takeaways:**

- Godot 3 works on the Raspberry Pi 4 due to the support for GLES3
- A user named efornara got Godot 3.0.6/3.1 with GLES2 working on the Raspberry Pi 3 B+ trough compiling and other methods, see:
https://github.com/godotengine/godot/issues/2671#issuecomment-452682275 (The error from compiling was apparently caused due to an upstream bug, see: https://github.com/godotengine/godot/issues/2671#issuecomment-452685013)
https://github.com/godotengine/godot/issues/2671#issuecomment-453222752
- There's a repo (https://github.com/efornara/frt) which let's the Pi support Godot 2.1/3.0/3.1 ~~BUT not~~ /3.2

**Other interesting things I found:**

- This website: http://raspbian.raspberrypi.org/raspbian/pool/main/g/godot/, which apparently has Godot 3.2 executables for the Pi
- On https://repology.org/project/godot/versions, there's also apparently Godot 3.2 listed as an official raspbian package, it's not marked _Stable_ but as _Testing_.
- When you want to install Godot trough `apt` on the Pi you write `apt install godot`, Godot 3.0.6 gets installed, as it's apparently the stable package (see link above).
- Someone got export templates for Pi working on 2.1.4: https://github.com/crabcrabcam/godot/releases
- This post: https://www.reddit.com/r/godot/comments/af1cji/compiling_godot_31_for_raspberry_pi/, where someone made instructions, on how to compile Gdot 3.1.1 for the Pi

**Describe the project you are working on:**
Applies to any project made on a Pi.

**Describe the problem or limitation you are having in your project:**
It's hard perhaps even impossible, to get Godot 3.2 working on the Pi.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It's hard to tell why Godot 3.2 doesn't work with Pis below 4. My assumption is that it has to do something with the missing GLES3 or because the Pi has an ARM processor (https://godotengine.org/qa/13114/export-godot-project-for-raspberry-pi). I think it could help to look into: https://github.com/efornara/frt

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
As I don't know the exact problems, I can't tell.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
A few lines of code probably won't do it, but Godot 3.2 is desired from Raspberry Pi users.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It isn't possible.

**Comment by efornara** [comment]
As a clarification, FRT (export template) does support Godot 3.2. In fact, I have just noticed this proposal when publishing 3.2.2-beta4 (https://sourceforge.net/projects/frt/files/previews/). Godot 3.2.1-stable is still available (it is the default download).

The unmodified x11 platform of Godot 3.2 should work out of the box on a Pi 4 and, by changing the OpenGL system settings, on a Pi 2/3 too. You should add a compiler flag to select hard float when generating the binaries, and I think some people have problems when using GCC, but, apart from that, it should be fine.

**Comment by asheraryam** [comment]
I've worked with godot 3.2 and 3.1 on the raspberry PI 3A and 3B+ and it works ok.

However, even for 2D there are a lot of optimizations and graphical limitations that the user needs to do and figure out on their own.

For example particles do not work at all, and the texture limit means that it's better to load things as needed instead of loading everything into the scene at once.

I still think it would be useful to make an official RPI export template, but this can only happen after https://github.com/godotengine/godot-proposals/issues/877 is decided since RPI is considered a low-end platform.

**Comment by antv199** [comment]
I found this [Vulkan update for the Raspberry Pi.](https://www.raspberrypi.org/blog/vulkan-update-now-with-added-source-code/) This might come in handy.

**Comment by Coolcreeper27** [comment]
This would help for beginning game devs to be able code any time

**Comment by Calinou** [comment]
@Coolcreeper27 
Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by dh928** [comment]
> I've worked with godot 3.2 and 3.1 on the raspberry PI 3A and 3B+
> 
> It works pretty well for 2D.
> 
> However, even for 2D there are a lot of optimizations and graphical limitations that the user needs to do and figure out on their own.
> 
> I still think it would be useful to make an official RPI export template, and I would love to help to make it happen since it would be useful to a lot of users, but this can only happen after #877 is decided since RPI is considered a low-end platform.
> 
> Example limitations: Particles do not work at all, and there's a texture memory limit that means that it's better to load things as needed instead of loading everything into the scene at once (or else they will appear black ingame).

How did you get it to work on your pi 3B+?

**Comment by asheraryam** [comment]
@dh928 You can use the [FRT thing](https://github.com/efornara/frt) directly or use [this thing](https://github.com/hiulit/RetroPie-Godot-Game-Engine-Emulator) which makes it run within RetroPie.

Note that some things are not supported, like particle effects or Websockets (but fortunately I used the other TCP option for multiplayer).

**Comment by ShalokShalom** [comment]
Would be nice to find it [here](https://archlinuxarm.org/) 🥳

**Comment by Calinou** [comment]
@ShalokShalom This can be handled by the Arch Linux ARM packagers already.

**Comment by ShalokShalom** [comment]
Yeah, but getting in touch with the distros and actually coming up with a ready to go build script would not hurt the visibility of the project. In the end, its all about making the software available to the people. I could do that 😊

**Comment by Edwardphantom** [comment]
> Yeah, but getting in touch with the distros and actually coming up with a ready to go build script would not hurt the visibility of the project. In the end, its all about making the software available to the people. I could do that 😊

man, plz, its gonna be amazing !

**Comment by nobody5050** [comment]
Bumping because this would be an awesome thing for godot to do and would also bring more users to the engine.

**Comment by Calinou** [comment]
@nobody5050 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by hiulit** [comment]
Hi everybody!

I'm writing from a Raspberry Pi 4 4GB where I successfully compiled all major versions (`2.1.6`, `3.1.0`, `3.1.1`, `3.1.2`, `3.2.0`, `3.2.1`, `3.2.2`, `3.2.3` and also `3.2.4 beta 4`) of Godot , both the editor and the export templates :tada: 

I could't compile `3.0.6` because it doesn't support GLES3. I'm not sure if that's the cause, but I couldn't anyway.

I want to clarify that I'm not an expert at compilattion, architectures and all that stuff, but I'm tech-savy enough to understand them and tinker with them.

I'll try to explain how I did it, what is working and what is not.

But first of all, I want to credit all the sources where I got the information from:

- @efornara 's [FRT project](https://github.com/efornara/frt/blob/master/doc/Compile.md).
- @w84death 's ["Raspberry Pi 4 as Perfect Indie Console"](https://bits.p1x.in/raspberry-pi-4-as-perfect-indie-console/) blog post.
- [Raspberry Pi GCC Toolchains](https://sourceforge.net/projects/raspberry-pi-cross-compilers/files/Raspberry%20Pi%20GCC%20Cross-Compiler%20Toolchains/Buster/GCC%209.3.0/Raspberry%20Pi%203A%2B%2C%203B%2B%2C%204/).

## How I compiled Godot (X11) for the Raspberry Pi

- I followed the official documentation about [Compiling for X11 (Linux, *BSD)](https://docs.godotengine.org/en/3.2/development/compiling/compiling_for_x11.html) to install all the dependencies (Debian/ubuntu).
- Then for the scons parameters and flags I used:
  - @efornara 's: `scons platform=x11 target=release_debug tools=yes use_llvm=yes CCFLAGS="-mcpu=cortex-a7 -mfpu=neon-vfpv4 -mfloat-abi=hard -mlittle-endian -munaligned-access" -j4`.
  - @w84death 's: `scons platform=x11 target=release_debug tools=yes use_llvm=yes CCFLAGS="-mtune=cortex-a72 -mcpu=cortex-a72 -mfloat-abi=hard -mfpu=neon-fp-armv8" -j4`.
  - A mix of @efornara 's and @w84death 's: `scons platform=x11 target=release_debug tools=yes use_llvm=yes CCFLAGS="-mtune=cortex-a72 -mcpu=cortex-a72 -mfloat-abi=hard -mlittle-endian -munaligned-access -mfpu=neon-fp-armv8" -j4`
  - The one from [Raspberry Pi GCC Toolchains](https://sourceforge.net/projects/raspberry-pi-cross-compilers/files/Raspberry%20Pi%20GCC%20Cross-Compiler%20Toolchains/Buster/GCC%209.3.0/Raspberry%20Pi%203A%2B%2C%203B%2B%2C%204/): `scons platform=x11 target=release_debug tools=yes use_llvm=yes CCFLAGS="-march=armv8-a -mfloat-abi=hard -mfpu=neon-fp-armv8" -j4`

For all I can tell, they all seem to work the same.

I don't know the differences between all those flags, but I found the [GCC ARM Options](https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html) site where all those options all explained.

As you can see, all the binaries are compiled using Clang.
GCC is suposed to be better (?), so I also compiled a version using GCC (without `lto`) to compare it with the Clang version and it seems to performe the same as the Clang binaries.
I could't compile using GCC with `lto` because it's very RAM consuming and apparently my Raspberry Pi run out RAM.

## Editor (IDE)

The only versions that work perfectly are `2.1.6` and `3.2.4 beta 4`.

I'm guessing that ` 2.1.6` works well because it's less demanding (?).
And for `3.2.4 beta 4`, it seems that it woks great because of the amazing job the Godot team has done with the GLES2 batching.

All the versions inbetween are very laggy, at least when run at fullscreen (1920x1200). If the editor window is smaller, at between ~1/2 and ~1/3 of that resolution, it performs correctly. I guess if all the new batching features were ported back to those versions they would work perfectly as well.

## Export templates

All versions seem to work fine.

To compile export templates

**Comment by Calinou** [comment]
> I could't compile using GCC with lto because it's very RAM consuming and apparently my Raspberry Pi run out RAM.

You'll need a Raspberry Pi with 8 GB of RAM and lots of patience for that, or cross-compile the ARM binary from an x86 machine.

**Comment by hiulit** [comment]
> > I could't compile using GCC with lto because it's very RAM consuming and apparently my Raspberry Pi run out RAM.
> 
> To compile an LTO-enabled build, you'll need a Raspberry Pi with 8 GB of RAM and lots of patience, or cross-compile the ARM binary from another machine with at least 8 GB of RAM.

Do you think that compiling with lto we would see noticeable improvements, performance-wise? Is it worth trying?

**Comment by Calinou** [comment]
> Do you think that compiling with lto we would see noticeable improvements, performance-wise? Is it worth trying?

On x86, it makes the binary 20-25% smaller and about 20-25% faster as well. I think it's very much worth it, but the performance gains may be different on ARM.

**Comment by efornara** [comment]
The official export templates (x11) work well on a Raspberry Pi 3 too. Today. Latest stable Godot release.

I am not complaining too much about a few additional stars and downloads that FRT is getting, but, really, most people who download FRT would be better served by downloading an official yet unsupported export template for Raspberry Pi 3/4.

In my opinion, at this point, FRT should become a niche "port", downloaded mostly by people targeting Raspberry Pi Zero, Raspberry 2/3 with Legacy driver or platforms where you have to use KMS/DRM (and possibly Wayland, if I find some time to add it).

I think that some sort of official support should be seriously considered for a future Godot 3.2 release, ideally 3.2.4-stable.

**Comment by Calinou** [comment]
> I think that some sort of official support should be seriously considered for a future Godot 3.2 release, ideally 3.2.4-stable.

We want it to happen, but I don't control the release process itself.

**Comment by efornara** [comment]
I suppose someone could publish unofficial templates for Godot, as occasionally happens. From what I have seen, these tend not to remain up-to-date for long. My guess is for lack of discoverability, which leads to lack of downloads / stars / likes, and ultimately to loss of interest from the person attempting it.

At the end of the day, I can live without an official export template. What I am missing is a reason, I suppose by who controls the release process itself, because, as things are now, the situation doesn't make much sense to me.

**Comment by Calinou** [comment]
> What I am missing is a reason, I suppose by who controls the release process itself, because, as things are now, the situation doesn't make much sense to me.

The [build containers](https://github.com/godotengine/build-containers) need to be updated to allow for ARM cross-compilation. It requires some additional work to do.

**Comment by hiulit** [comment]
@efornara Do you have any tips on which flags are best suited for compiliing for both the Raspberry Pi 3 and 4?

As you can see in my post above I used diferrent flags, but as far as I could see, they all seemed to perform equally.

**Comment by efornara** [comment]
@Calinou 

> The build containers need to be updated to allow for ARM cross-compilation. While feasible, it's not 100% trivial.

Fair enough. I am planning to soon focus on cross-compiling for FRT too. I did a quick test a few months ago, and I think I only had to make just a few minor changes to my detect.py to get a binary that started without crushing.

@hiulit 

> Do you have any tips on which flags are best suited for compiliing for both the Raspberry Pi 3 and 4?

Not really. I copied the flags that I am using from an unchallenged forum post on the Raspberry Pi forums ages ago. Today, I would use something different (i.e. less specific). I have never changed them because it wasn't broken, so there was no point in risk breaking it. I do plan to simplify them when I switch to cross compiling: I am making a break anyway, so I might as well go all the way.

Same boat as you when it comes to performance. I tried once a benchmark and the results were inconclusive and too noisy.

About LTO. I had problems generating a Godot 3.something editor build, even on a PC with 8G of RAM running qemu-user-static overnight. There is a possibility that a 8G Pi4 won't be enough to generate a 32-bit binary, and you might need to cross compile anyway.

**Comment by hiulit** [comment]
Hi everyone!
In case anyone is interested, I've just released **[Unofficial Godot Engine for the Raspberry Pi](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi)**: Unofficial Godot Engine editor and export templates binaries for the Raspberry Pi.

**Comment by Edwardphantom** [comment]
> Hi everyone!
> In case anyone is interested, I've just released **[Unofficial Godot Engine for the Raspberry Pi](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi)**: Unofficial Godot Engine editor and export templates binaries for the Raspberry Pi.

You are a hero bro, i wanna be like you when i grow up.

**Comment by alexfreyre** [comment]
Nice! personally I feel extremely happy because of this PR :)

I am a visual artist from the field of fine arts and I use Godot for what we call "new media art" so, I do applications as  interactive virtual experiences, for that reason I don't make games nor publish anything like that, 

I do exhibitions and I need to get them running on galleries, so SBC always have been the 1st option for me because its low cost. That said, for me too is very important this arm support, but not only for RPI, I explain why:

FRT project is amazing because it runs on almost a raw archlinux arm system from command-line and the performance is much better than from X11, but it is focused on RPIs, which is really nice! they are the most known SBC but not the cheapest nor powerful either.

For other boards we need to compile and run from X11 which is considerably slow while those boards are much more powerful than a RPIs, for example Odroid N2+ with a 850MHz GPU at $63, I have found other which I can't remember the brand nor the processor but it cost $19 with a 650MHz GPU... in my humble opinion it worth support a wide scenario (I am not a developer I can't imagine the time and the effort to do that) because arm powered custom devices/creations can be very useful/cheap and widely used from art piceces, design, custom gaming-machines, and who knows!!! 

So my point/wishes/requests are:
1. Please support in general arm and aarch64 (being arm6 useful on 7 and 8), would be great, because there are a lot of boards on the market with a lot of different processors
2. if it is possible, please add the wonderful feature to run godot releases from command-line like FRT but not only limited for RPIs or the processors they have. (if this involve a complex workflow with several drivers or something like that, consider support a few of the most known GPUs)

amazing all the people concerned about this!!! (sorry about my english and this looong message :) )

**Comment by silverkorn** [comment]
@alexfreyre , kinda same here. 
I'm trying to show a friend in this field that they can use Godot for artistic exposition at low price and few knowledge (They were taught Python in their courses)

**Comment by efornara** [comment]
I hope the moderators don't mind if I comment about FRT here.

@alexfreyre

FRT isn't limited to Raspberry Pi. Here is my own experience with an Orange Pi PC (AllWinner H3 with Mali 400).

I started with a "server" buster image from Armbian and compiled a newer version of Mesa myself. The official FRT binaries work both on the console (KMS/DRM) and on X11. The Godot 2.1 2D platformer runs at 30 fps on the console and at 60 fps on X11.

My godot2-editor from https://github.com/efornara/godot2-rpi-repo (basically, Godot 2.1 patched to use EGL/ES2 instead of GLX/GL2) works too. I wouldn’t say that it is perfectly smooth, but I have found it quite useable, much more than a 3.2.4beta4 editor on a Pi 3B+.

I did try once a "desktop" buster image from Armbian, and I didn’t get any hardware acceleration, so that might have been the problem you got. I suggest to check es2_info and es2gears to make sure that it really is a Godot / FRT problem and not an OS problem. IIRC, I got a "DRI2: failed to authenticate" as warning that there was no hardware acceleration.

For anyone looking for cheap boards, I also tried the Godot 2.1 2D platformer (console) on a Orange Pi One (same as the PC, but with fewer ports and only 512M of RAM). Again, it works fine, but, to be fair, having to compile Mesa yourself might not be for everyone.

Also, see: https://github.com/efornara/frt/issues/25

**Comment by alexfreyre** [comment]
@efornara , I  hope that  in some moment Godot supports natively EGL, and if this PR is considered finally that the export templates could support EGL (only if performance is better)

but, how can platformer 2D runs at 30 fps on the console and at 60 fps on X11? isn't EGL faster than X11? x11 need some hardware resources..

**Comment by efornara** [comment]
Unless you are using a composer, or you are filling up your memory with heavy X11 applications, the overhead of using X11 instead of KMS/DRM should be negligible. I did my test with only openbox and xterm running. Pi 0-3 + legacy driver is a special case with different trade-offs.

I didn't investigate why KMS/DRM was slower. One explanation could be that the platformer size was smaller than the display. While on KMS/DRM the game "window" was the same size as on X11, maybe the size of actual framebuffer used was as big as the full display (1024x600 in my case). There might be some possible optimization to be made on the FRT side. When/if I put my hands on the kmsdrm module I might have a look.

**Comment by xarses** [comment]
I'm happy that you have gone on about FRT and other alternatives to official support here, as this helps lay the path for people that want understand how to do this themselves

**Comment by samdze** [comment]
I recently managed to build Godot 3.2.3 export templates on a Allwinner R16 (A33), the cpu contained in a portable device called [GameShell](https://www.clockworkpi.com/gameshell).
Not the easiest task, but in the end it turned out very well for 2D games, the only issue currently is the audio, which is a bit distorted but mostly tolerable.

I was also able to compile a release build using LTO even if the machine has only 1GB of ram.
It should be possible to do the same thing on a Raspberry allocating a large enough swap area/file.

Here I described the complete build process: https://github.com/samdze/godot-gameshell

Godot running on a device like that makes me very happy and I'd love to get official support for at least a few very popular arm devices. (Raspberry Pi, for sure)

**Comment by hiulit** [comment]
Great work @samdze !

For the audio issues, this might help you https://github.com/godotengine/godot/pull/43928

**Comment by nicolas-van** [comment]
Just to say I would really be happy to have Godot working on a Raspberry pi too. Continue like that guys 😎

**Comment by Calinou** [comment]
@nicolas-van Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by hiulit** [comment]
If anyone is interested, I've added **headless** and **server** builds https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi.

I couldn't compile them for 2.1.6 for whatever reason.

I don't have any experience with that kind of builds, so if would be great if someone could test that they work :)

**Comment by JoaoSilverio255** [comment]
Hey, I don't know if the discussion is still active or not but I have tried to make it work and for some reason, the change_scene() method doesn't seem to work. Have anyone came across this error?

**Comment by Calinou** [comment]
@JoaoSilverio255 Please report issues related to unofficial builds to their author *after* making sure you can't reproduce them on official builds.

**Comment by AgentSmith0** [comment]
Is it possible to export a project for PinePhone devices, that run on Linux (ARM)? https://wiki.pine64.org/wiki/PinePhone

**Comment by cumminsd** [comment]
Like the "new media art" posts above, another non-game use case is for the small, touch screen, home automation devices. Using Godot to program the displays would be awesome. :-)

**Comment by xarses** [comment]
totally off topic at this point, but I'm working on a control app that was initially intended for use by streamers that I'm planning on using for that too, and someone asked for the support for ARM for similar reasons as yourself here (which is why I'm following the arm support thread) Ping me on discord if you want to chat about it

**Comment by efornara** [comment]
I think I mentioned cross-compiling in this thread.

In the latest versions of FRT, I did finally switch to cross compiling my binary releases (at least for arm32v7 and arm64v8, the architectures everyone is interested in anyway).

Quick start is here: https://github.com/efornara/frt/blob/2.0.0/doc/Compile.md. My dockerfiles (image based on debian/buster) are here: https://github.com/efornara/crossbuild and an example of detect.py is here: https://github.com/efornara/frt/blob/1.1.0/detect.py. Tested with both gcc and clang. I don't normally use LTO, but, when I tested it, it only worked on one. IIRC, it was clang.

Maybe there are problems I am not seeing yet, but so far it seems to be working. To be honest, it was easier than I expected. I just select the proper triple for the compiler and use the triple- version of pkg-config.

**Comment by silverkorn** [comment]
About the recent proposal renaming, wasn't the request not explicitly only about the export template but also the editor?
If not, maybe create another proposal for the editor?

**Comment by hiulit** [comment]
Hey! For anyone interested, I've added 64 bits binaries for the Pi 4 https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi

**Comment by hiulit** [comment]
Also, I've created a PPA repository so you can easily install Godot using `apt install godot` https://github.com/hiulit/ppa

**Comment by aaronfranke** [comment]
Related to official ARM builds, I've been working on improving support for non-x86 architectures in Godot. Currently Godot's build system, export system, GDNative, and more have poor support.

The build system has a mix of `bits` and `arch` and `android_arch`, but it really should be unified into one architecture argument. I have a PR for this here https://github.com/godotengine/godot/pull/55778 and I also have another PR to do this in GDNative C++ in 3.x here https://github.com/godotengine/godot-cpp/pull/714 and in the GDNative C demos here https://github.com/godotengine/gdnative-demos/pull/63

The export system in 3.x was missing a dedicated class for LinuxBSD and instead used the "PC" class for it. In Godot 4.0 this has been fixed by @bruvzg, now there is an EditorExportPlatformLinuxBSD class https://github.com/godotengine/godot/pull/58455 . Because of this, we were able to update the logic for the export template file names to be less hard-coded for x86 https://github.com/godotengine/godot/pull/59416 . The export system also has a `binary_format/64_bits` flag instead of architecture selection and you need to specify custom templates to do an ARM export, this needs to be replaced too.

When loading a PCK file, currently it looks for the executable name, and then the executable name with the extension stripped. This system works for one executable, and works for multiple executables on Linux (like `mygame.x86_64` and `mygame.arm64` loads `mygame.pck`) but we currently can't have multiple executables load the same PCK on Windows. I have a PR to improve the PCK loading logic so that we can have `mygame.x86_64.exe` and `mygame.arm64.exe` etc load `mygame.pck` https://github.com/godotengine/godot/pull/59527

**Comment by Calinou** [comment]
We discussed this in a proposal review meeting. We largely agreed to the feature and it has significant community support. Some more comments:

- Having a X11 backend only is fine (at least for now), since the Raspberry Pi 4 handles this quite well.
- There are plans to provide official builds, but the Linux distribution compatibility still needs to be figured out. We need to make sure our x86 build container cross-compiles binaries that work on older Linux ARM distributions, including the Raspbian version typically used on Raspberry Pi 3 and 4. Both ARMv7 and ARMv8 (64-bit) binaries should be provided to maximize compatibility, as Raspbian was 32-bit by default until early 2022.
- The builds will be made to work on various ARM SBCs that support OpenGL (ES), not just Raspberry Pi. It should even be possible to run headless builds on Android devices using Termux :slightly_smiling_face: 
- The Vulkan Clustered backend is unlikely to work properly on Raspberry Pi 4, but Vulkan Mobile backend might work. Still, OpenGL should be preferred as it has a lower base cost on simple scenes (which will typically be used on low power devices like Raspberry Pi).
- Some modules such as navigation may be disabled at first, as they don't build for Linux ARM correctly out of the box. This will probably be resolved in the future.

**Comment by Faless** [comment]
For those interested, I managed to get `3.x` build cross compiled from debian for raspbian.
This process is much faster then the usual compilation via qemu or on device (you can get an editor build in 10/15 minutes compiling with `-j5`).
I managed to produce both release builds and editor builds by only disabling few modules (haven't tested LTO yet), but the result is actually pretty great and does not require code changes to Godot (though, fixing the disabled modules likely will).

Here is the dockerfile (you can achieve the same on a true debian bullseye running the various apt commands ofc):

```
FROM debian:bullseye-slim

RUN set -eux; \
        dpkg --add-architecture armhf && \
        apt-get update && \
        apt-get install -y g++-arm-linux-gnueabihf scons pkg-config build-essential git && \
        apt-get install -y libstdc++-10-dev:armhf libx11-dev:armhf libxcursor-dev:armhf libxinerama-dev:armhf libgl1-mesa-dev:armhf libglu-dev:armhf libasound2-dev:armhf libpulse-dev:armhf libudev-dev:armhf libxi-dev:armhf libxrandr-dev:armhf && \
        rm -rf /var/lib/apt/lists/*

CMD /bin/bash
```

From the image, you can then build Godot.

```bash
# Build the image
docker build -t pibuilder:latest -f Dockerfile.pi .
# Run a container with the built image
docker run -it --rm pibuilder:latest /bin/bash
# Checkout godot (only 3.x branch tested)
git clone https://github.com/godotengine/godot.git --depth 1 -b 3.x
# Cross compile the editor
PKG_CONFIG_PATH=/usr/lib/arm-linux-gnueabihf/pkgconfig/ scons CC=arm-linux-gnueabihf-gcc CXX=arm-linux-gnueabihf-g++ LINK=arm-linux-gnueabihf-g++ AR=arm-linux-gnueabihf-ar platform=x11 target=release_debug arch=armhf module_denoise_enabled=no module_theora_enabled=no module_webm_enabled=no CCFLAGS=-Wno-attributes use_static_cpp=no module_raycast_enabled=no
```

The produced binary was tested working on both the PI3 and the PI4.

As you see, this did not require changes to godot itself, since our build system is pretty flexible and allows heavy customization via flags.
I suspect we could achieve a similar result on the base fedora image with use for most builds (probably not the ubuntu one which has a toolchain from 5+ years ago). The choice of debian was mostly for two reason:
- I'm an `apt` aficionado :-p
- Raspbian is based on debian, which guarantees library version compatibility since as mentioned by Calinou distro compatibility in the linux world (especially in the ARM/SoC world) is still tricky.

I hope this helps others that are trying to build for PI while we figure out how to properly integrate it in our release cycle.
I could go ahead and open a PR on the build-containers/build-scripts repo but we probably want to iterate on this, so I'll drop a ping to @akien-mga and @hpvb for some feedback.

**Comment by hiulit** [comment]
Thank you very much @Faless!!
I've successfully compiled a 32 bits binary of the editor. I'll try it on my Pi and I'll report back.
Could this image be used to compile 64 bits binaries? If so, how?

P.S. You are missing a `cd godot` after cloning Godot ;)

**Comment by hiulit** [comment]
Ok, so I've tested it on my Pi 4 and it didn't work, but because I'm still using Buster and the image uses Bullseye :)

FYI: I tried to create a new image with and old distro to have better compatibility and the errors started to pop up:

- Ubuntu 14.04: Failed to fetch armhf packages.
- Ubuntu 16.04: Failed to fetch armhf packages.
- Debian Jessie: Unable to locate package `g++-arm-linux-gnueabihf`.
- Debian Stretch:  Unable to locate package `libstdc++-10-dev:armhf`.
- Debian Stretch (without `libstdc++-10-dev:armhf`): scons: building terminated because of errors.
- Debian Buster: Unable to locate package `libstdc++-10-dev:armhf`.
- Debian Buster (without `libstdc++-10-dev:armhf`): **OK**.

**Comment by Faless** [comment]
I found out it is also possible to build for Raspberry PI 3/4 using the [godot buildroot toolchain](https://download.tuxfamily.org/godotengine/toolchains/linux/).
See instructions to ["install"](https://github.com/godotengine/buildroot#using-the-sdks)

Godot 4.0:

```bash
export TOOLCHAIN_PATH=/path/to/godot/toolchain
PATH=${TOOLCHAIN_PATH}:${PATH} scons platform=linuxbsd target=release_debug arch=armhf CCFLAGS="-mfpu=neon -mfloat-abi=hard -mlittle-endian" module_raycast_enabled=no module_etcpak_enabled=no module_denoise_enabled=no module_theora_enabled=no -j9
```

Godot 3.x:

```bash
export TOOLCHAIN_PATH=/path/to/godot/toolchain
PATH=${TOOLCHAIN_PATH}:${PATH} scons platform=linuxbsd target=release_debug arch=armhf CCFLAGS="-mfpu=neon -mfloat-abi=hard -mlittle-endian" module_raycast_enabled=no module_denoise_enabled=no module_theora_enabled=no module_webm_enabled=no -j9
```

The produced builds should likely be more compatible then the ones from the debian toolchain.

**Comment by hiulit** [comment]
Yeah, that's what I'm using for my [unofficial Godot binaries for the Raspberry Pi](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi). It works great but it doesn't have support for 64 bits, so those I have to compile on the Pi itself.
I've created a [script to easily cross-compile](https://github.com/hiulit/cross-compile-godot-raspberry-pi) from a Linux machine.

**Comment by abczezeze** [comment]
> Yeah, that's what I'm using for my [unofficial Godot binaries for the Raspberry Pi](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi). It works great but it doesn't have support for 64 bits, so those I have to compile on the Pi itself. I've created a [script to easily cross-compile](https://github.com/hiulit/cross-compile-godot-raspberry-pi) from a Linux machine.

The ```Unofficial-Godot-Engine-Raspberry-Pi``` repo. It support raspbian 64 bit. I've used this repo.
[purchase](https://th.cytron.io/p-raspberry-pi-4-8GB-starter-kit), [install](https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-64-bit
), [post](https://twitter.com/abczezeze/status/1517805080401289216?cxt=HHwWgIDR-bnrqZAqAAAA)

**Comment by FeralBytes** [comment]
FYI, using the Raspberry Pi 64 ARM templates I was able to export to my PinePhone Pro. But the device display size reported by Godot is 360(W)x720(H) in portrait. Even though it is actually 720x1440 in portrait.

**Comment by skysphr** [comment]
I can confirm, the 64 bit export template also works out of the box with (standard) PinePhone running Manjaro ARM.

**Comment by aaronfranke** [comment]
(continuing from [my last comment](https://github.com/godotengine/godot-proposals/issues/988#issuecomment-1096890844))

I did some work on the export logic to support multiple architectures. I made this PR https://github.com/godotengine/godot/pull/63096 for renaming the existing export template file names from `.64` to `.x86_64` etc. With all that done, I made this PR https://github.com/godotengine/godot/pull/63563 to replace `binary_format/64_bits` with a more general `binary_format/architecture`. You can supply export templates for different architectures by dropping them in `~/.local/share/godot/export_templates`. The export dialog's architecture selection now looks like this:

<details>
<img src="https://user-images.githubusercontent.com/1646875/181395798-898dfc45-cee6-4e53-9e6d-c5c0c1fb1c83.png" data-canonical-src="https://user-images.githubusercontent.com/1646875/181395798-898dfc45-cee6-4e53-9e6d-c5c0c1fb1c83.png" width="450" height="451" />
</details>

Related to the above, I updated the export dialog to allow exporting a PCK without export templates installed. This way, if you only have an editor but no export templates, you can test a game export on your Raspberry Pi with a PCK plus the editor. https://github.com/godotengine/godot/pull/62925

My PR for unifying the `bits`, `arch`, and `android_arch` arguments https://github.com/godotengine/godot/pull/55778 has been merged. It did have some regressions, particularly on Windows and Android, special thanks to @akien-mga @bruvzg @m4gr3d @neikeq @SaracenOne for fixing issues with it so quickly. With this PR merged, the arch handling code is a lot cleaner, and when compiling on a Raspberry Pi it should auto-detect ARM, so you don't need any extra arguments.

I also have this small PR https://github.com/godotengine/godot/pull/64366 which makes architecture compile options on Linux more explicit. Having this might help with cross-compiling, though I haven't tested cross compiling. Also worth mentioning, if you want to target a specific hardware feature set, you can modify the build system here, for example as far as I know this option `-march=armv8-a+crc+simd -mtune=cortex-a72` will optimize for the Raspberry Pi 4's CPU specifically.

**Comment by 37Rb** [comment]
I'd love to see both editor and export templates for Raspberry Pi but official export templates are more important than the editor. I suspect the most common use case would be developing on a larger computer and exporting to a Pi being used like a game console. Consider making the official export template work with any ARM Linux board that has the required graphics APIs, not specific to Raspberry Pi. There are Raspberry Pi competitors and even clones that some people might use.

**Comment by Calinou** [comment]
> Consider making the official export template work with any ARM Linux board that has the required graphics APIs, not specific to Raspberry Pi.

This should already be the case, as long as the binary is compiled against an old enough glibc to work on many ARM Linux distributions. ARM64 will likely be required for performance reasons though (the performance difference between ARMv7 and ARM64 is quite significant).

**Comment by aaronfranke** [comment]
There are a lot of use cases for ARM Linux support, including Raspberry Pi, ARM Linux [free servers](https://www.servethehome.com/oracle-cloud-giving-away-ampere-arm-a1-instances-always-free/), ARM Linux on Apple Silicon, ARM Linux on Chromebooks, Pine Phone, other Pine devices, and many other ARM Linux devices. I agree that it would be best for Godot to support ARM Linux in general, not just the Raspberry Pi.

**Comment by ShalokShalom** [comment]
I guess Android and iOS tablets?

**Comment by Calinou** [comment]
> I guess Android tablets?

Android tablets can't run Linux binaries (outside of Termux, but it's limited to CLI apps only).

**Comment by 37Rb** [comment]
> This should already be the case, as long as the binary is compiled against an old enough glibc to work on many ARM Linux distributions.

You're right. I ended up getting an [Orange Pi 4](http://www.orangepi.org/html/hardWare/computerAndMicrocontrollers/details/orange-pi-4-LTS.html) because Raspberry Pi 4s are on back order. Installed Ubuntu 22.04 on it and Godot executables made using the [unofficial export template](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi#351) work great on it! I was pleasantly surprised, thought I would run into at least one problem.

**Comment by mots** [comment]
I think ARM Linux support would also allow for easy ports to various handheld devices like the [Anbernic rg353v](https://anbernic.com/products/rg353v-rg353vs).

**Comment by hiulit** [comment]
For those of you who don't know, @efornara has stopped maintaining FRT :(
https://github.com/efornara/frt

I hope official support comes in the near future.

**Comment by ShalokShalom** [comment]
> I think ARM Linux support would also allow for easy ports to various handheld devices like the [Anbernic rg353v](https://anbernic.com/products/rg353v-rg353vs).

Well, that device already supports Android.

**Comment by orowith2os** [comment]
I'd like to note that Godot would only have to support aarch64 Linux; aarch32 Linux is not required, and is, in fact, no longer supported on anything newer than `org.freedesktop.Platform//20.08` - aka the past three years. Even native x86 support was dropped in favor of *only* x86_64 with (optimized for x86_64) i386 compat libraries. 

This would be nice, as Apple Silicon is getting a Linux port, and it already provides the [first GPU diver for Apple Silicon to be fully conformant](https://rosenzweig.io/blog/first-conformant-m1-gpu-driver.html). We at [Flathub](https://flathub.org) would also appreciate it if:
1. Godot applications can be built from source in an offline environment
2. Godot applications supported aarch64 Linux

**Comment by bruvzg** [comment]
Generic ARM Linux (both 32 and 64-bit) is already supported, but it's using desktop GLX. So [EGL driver](https://github.com/godotengine/godot/pull/68944) (and optionally GL renderer changes added in [ANGLE PR](https://github.com/godotengine/godot/pull/72831), to support loading both desktop GLX and EGL/GLES in runtime) should be the only thing missing, or is there something else?

**Comment by orowith2os** [comment]
An EGL driver, like you mentioned, sounds like enough? It would also be necessary for Wayland, so two birds with one stone. Getting everyone off of GLX is a sound move, at least.

I'm not sure what the problem with using GLX here is, though; what makes it desktop-specific? Is compiling it for aarch64 not enough?

**Comment by bruvzg** [comment]
> I'm not sure what the problem with using GLX here is, though; what makes it desktop-specific?

Are there drivers for desktop OpenGL/GLX on Raspberry Pi and similar devices? I have not used any of these for non-headless application for years, so I'm not sure about current status, but I think it always was GLES via EGL only.

**Comment by orowith2os** [comment]
There's a Vulkan driver now, so if desktop GL isn't available, it can fall back to Zink (GL on Vulkan) if necessary. But it sounds like there's no hardware support for desktop GL, so it would indeed use that fallback.

**Comment by TheJackiMonster** [comment]
From my testing running Godot 3 (installed as flatpak via flathub.org) works great on the Librem 5 without issues. I could run my existing game on it easily as well. However the export template for Linux did build to x86, I think. So I couldn't export for ARM even though the game was running on ARM.

Since Linux phones like the Librem 5 or Pinephone are quite usable in desktop form factor as well, it's not unlikely people will try to develop on it (when docked of course since mobile form factor is not ideal for Godot currently ^^'). Having the option to export the game for ARM Linux out of the box seems very usable because of this.

I mean, this is a great benefit for developers that they are able to easily show anyone their game with a mobile device in action, get feedback and potentially immediately implement a change or test adjustments. I would argue it's similar to having the Steam Deck run Godot in handheld form factor and it makes the engine quite versatile. Honestly you don't even notice it's starting a full blown game engine on this small phone - no comparison to the loading screens with other competition. ^^'

**Comment by AlexHCC** [comment]
Another use case for Linux ARM export, is headless servers for multiplayer games.
ARM servers are becoming more and more popular, and are usually hosting plans are a bit cheaper compared to x86.

**Comment by Calinou** [comment]
> I see the Architecture field on the export dialog window, does that mean the arm32 and arm64 are compatible with single board computers like RPi and Orange Pi?

Yes, but you need to compile export templates for it.

**Comment by fbcosentino** [comment]
In other words, Godot 4 requires Raspberry Pi 4 and similar level..?

**Comment by Calinou** [comment]
> In other words, Godot 4 requires Raspberry Pi 4 or similar level..?

Yes, otherwise you may run out of RAM and have a generally low framerate, especially if trying to to use the editor.

**Comment by BenMcLean** [comment]
> On top of that, do 64 Bit applications and operating systems use a bit more RAM as well. I think a Raspberry Pi 4 as a minimal hardware requirement is appropriate.

Wait that means we couldn't use Godot to program a GUI for a Raspberry PI Zero app?

I guess Godot has to decide what it wants to be and not every tool is for every application. But some of the more basic GUI stuff Godot can do seems like it would be something you'd want to be able to run on Raspberry PI Zero.

**Comment by aaronfranke** [comment]
For using Godot as a lightweight GUI toolkit, see proposal #190. I have several open pull requests towards this goal that are in need of reviews. Also note that in general export templates will be lighter than the editor.

**Comment by ecurtz** [comment]
> On top of that, do 64 Bit applications and operating systems use a bit more RAM as well. I think a Raspberry Pi 4 as a minimal hardware requirement is appropriate.
> 
> But what would be needed, to close this issue successfully?

The editor will build on the Pi4 with Vulcan after a couple of hacks to the Vulcan pool allocator as described here. https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi/issues/29#issuecomment-1499520481 My impression is these are implementation specific and the better solution would be to check the response code when querying for the pool ID (this is from memory, some of the terms may be wrong.) I haven't tested it much but the editor builds and runs. Previous versions of that unofficial build repository worked on earlier Pi hardware but that was all OpenGL, and I'm nowhere near familiar enough with Godot internals to know how difficult it would be for those to work with 4.x.

**Comment by AwokeKnowing** [comment]
I'm using OrinNX for some robotics and considering using godot for some visualization.  plenty ram and compute, but what exactly do I need to do to compile/use godot editor and run things on Orin NX?

**Comment by liberodark** [comment]
I hope you can add support to ARM.

**Comment by Calinou** [comment]
@liberodark Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by hiulit** [comment]
It seems that we are finally getting official ARM binaries :D https://x.com/Akien/status/1719691269910282698

You can download them here https://downloads.tuxfamily.org/godotengine/testing/4.2-beta-2023-11-01-linux-arm/

**Comment by hiulit** [comment]
I've made some tests on the latest Pi 5 and sadly, there seems to be some issues with the Vulkan drivers on Raspberry Pi OS.
But the good news are that they are working on Ubuntu 23.10 :)

Here's a video running the 2D platformer demo on the Pi 5 on Ubuntu 23.10 using `--rendering-driver opengl3_es` https://x.com/hiulit/status/1723825137030451599?s=20

**Comment by RadiantUwU** [comment]
Here's a suggestion: Temporarily fix this by providing debug and release templates for Linux arm64 headless only for servers and such

**Comment by Calinou** [comment]
> Here's a suggestion: Temporarily fix this by providing debug and release templates for Linux arm64 headless only for servers and such

[4.2.rc1](https://godotengine.org/article/release-candidate-godot-4-2-rc-1/) has official Linux ARM64 + ARM32 binaries for the editor and export templates, which can be run with `--headless` as usual.

**Comment by Netherite1701** [comment]
> I've made some tests on the latest Pi 5 and sadly, there seems to be some issues with the Vulkan drivers on Raspberry Pi OS.
> But the good news are that they are working on Ubuntu 23.10 :)
> 
> Here's a video running the 2D platformer demo on the Pi 5 on Ubuntu 23.10 using `--rendering-driver opengl3_es` https://x.com/hiulit/status/1723825137030451599?s=20

Did you test the ubuntu run on the RPI5? Or on another computer?

**Comment by hiulit** [comment]
> Did you test the ubuntu run on the RPI5? Or on another computer?

I tested Ubuntu 23.10 on the RPI5.

**Comment by OverloadedOrama** [comment]
Now that Godot 4.2 stable is out with official ARM builds, shouldn't this issue be closed? Or are there more things that yet remain to be implemented?

**Comment by TheJackiMonster** [comment]
> Now that Godot 4.2 stable is out with official ARM builds, shouldn't this issue be closed? Or are there more things that yet remain to be implemented?

Depends... originally the issue was about Godot 3 which still supports more devices with lower specs because of the OpenGL ES 2.0 renderer. For example I tried to arm64 build with compatibility profile on my Librem 5 and it couldn't run because it still required OpenGL 3.3 which the etnaviv driver on it does not support yet.

So I personally would still like to have ARM builds in Godot 3. I'm not sure whether there is still an update planned for Godot 3. But it would be neat.

**Comment by aaronfranke** [comment]
@TheJackiMonster It should already be possible to make Arm builds for Godot 3. Even though the code for it is a bit janky, and there are no plans to backport the architecture refactoring changes in Godot 4, it should still work.

**Comment by ecurtz** [comment]
The 4.2 release announcement said there were known issues with the Raspberry Pi build, but I didn't see them documented in a quick search. I wouldn't think this should be closed unless those are fixed or new, more focused issues are added so they can be tracked.

**Comment by fbcosentino** [comment]
Issues with Raspberry Pi build, or with Raspberry Pi OS?

**Comment by TheJackiMonster** [comment]
> @TheJackiMonster It should already be possible to make Arm builds for Godot 3. Even though the code for it is a bit janky, and there are no plans to backport the architecture refactoring changes in Godot 4, it should still work. I was able to port the RISC-V code to Godot 3 and it compiles/runs fine, so I would expect Arm to also work.

Do you have any instructions for me? Are we still talking about export templates or engine builds? Because if I run arm64 Godot 3, I still get the options to export to x86. Otherwise running a game through the editor works. But if there was an export template, I could use, that would solve my issue.

**Comment by ecurtz** [comment]
> Issues with Raspberry Pi build, or with Raspberry Pi OS?

"in fact, we’re already aware of issues running executables on Raspberry Pi OS". 
OS, apparently. If I just missed those and there are already new issues here for that then I don't mind this one being closed.

**Comment by hiulit** [comment]
> > @TheJackiMonster It should already be possible to make Arm builds for Godot 3. Even though the code for it is a bit janky, and there are no plans to backport the architecture refactoring changes in Godot 4, it should still work. I was able to port the RISC-V code to Godot 3 and it compiles/runs fine, so I would expect Arm to also work.
> 
> @aaronfranke Do you have any instructions for me? Are we still talking about export templates or engine builds? Because if I run arm64 Godot 3, I still get the options to export to x86. Otherwise running a game through the editor works. But if there was an export template, I could use, that would solve my issue.

You can use my [unofficial raspberry pi builds](https://github.com/hiulit/Unofficial-Godot-Engine-Raspberry-Pi).

**Comment by hiulit** [comment]
I don't know what are the godot team plans, but I would wait at least until you can download ARM binaries from the website. That is what "official" means to me.

**Comment by hiulit** [comment]
Here some interactions I had with Rémi and other people about this topic:

https://x.com/Akien/status/1719691269910282698?s=20

https://x.com/Akien/status/1727806022792712332?s=20

**Comment by akien-mga** [comment]
I'll repost what I wrote on the Godot Forum [a few days ago](https://forum.godotengine.org/t/godot-4-x-on-the-raspberry-pi/34417/4):

So I've done some testing on Raspberry Pi 5 8GB RAM indeed, all on RPi OS.

What I found out so far:

- Our new official Linux arm64 binaries in 4.2 (added to the official set of binaries a few weeks ago) don't work on RPi OS, they segfault. This seems caused by the toolchain we use for cross-compilation, I'll need help from @hpvb to fix it. I'm told those same binaries work on Ubuntu 23.10 on RPi 5, so it might be a RPi OS specific bug. Still something we need to fix as it's the base distro.

- Self compiled binaries on real arm64 hardware, or possibly cross-compiled with other toolchains (e.g. clang aarch64 target) should work fine however. So the segfault is really just a problem with how we do the builds, not with Godot itself.

- With a self-compiled build, I was able to test how things work. Both Vulkan and OpenGL ES seem to work ok. For Vulkan, note that RPi OS does not install Vulkan drivers out of the box, you have to install `mesa-vulkan-drivers`. Other distros might have similar oversights.

- I haven't had luck opening 3D projects with meshes like Kenney's FPS Starter Kit. I think there's a problem with VRAM compression of textures, and generally it seemed to take forever to import assets. I might have just given up too fast.

- It's easy to exhaust the resources of the RPi 5 and make RPi OS deadlock, possibly due to thermal throttling. Though my first tests were done before installing Active Cooling, now with the fan it seems to be doing ok.

- There's still a lot more to test and optimize, but for now I want to fix our official binaries.

---

I also want to provide official Linux ARM binaries for Godot 3.6, but it currently suffers the same issue that the binaries produced by [our Linux SDK](https://downloads.tuxfamily.org/godotengine/toolchains/linux/aarch64-godot-linux-gnu_sdk-buildroot.tar.bz2) seem to crash on Raspberry Pi OS. It's not a Godot problem, even a hello world C program crashes on start.

I made a few adjustments to the `3.x` branch last week to make sure it builds out of the box on arm64. Notably https://github.com/godotengine/godot/commit/dede8d384f6bfcf66bc0468cb34736a8c1ba63ad and #85915, so that embree, theora, and webm all build fine too (so no module needs to be disabled).

**Comment by akien-mga** [comment]
> * Our new official Linux arm64 binaries in 4.2 (added to the official set of binaries a few weeks ago) don't work on RPi OS, they segfault. This seems caused by the toolchain we use for cross-compilation, I'll need help from @hpvb to fix it. I'm told those same binaries work on Ubuntu 23.10 on RPi 5, so it might be a RPi OS specific bug. Still something we need to fix as it's the base distro.

Well that went fast, I got @hpvb remote access to my RPi 5 yesterday, and they fixed the issue in our Linux arm64 toolchain: https://github.com/godotengine/buildroot/commit/e6c2704eb22f0ea2a5dbe3d3084cec414ae540c8

We now have a new version of our Linux SDKs that should produce valid arm64 binaries for RPi OS: https://downloads.tuxfamily.org/godotengine/toolchains/linux/2023-12-11/

I'm using these new Linux SDKs for 4.2.1-stable, so it should work out of the box on arm64. Please test and confirm:

- https://downloads.tuxfamily.org/godotengine/4.2.1/Godot_v4.2.1-stable_linux.arm64.zip

I also made a test build from the `3.x` branch with the same SDKs, which also seems to work for me on both RPi 5 on RPi OS, and the Linux subsystem on an arm64 Chromebook:

- https://downloads.tuxfamily.org/godotengine/testing/Godot_v3.6-beta_x11_arm64.zip

**Comment by akien-mga** [comment]
So this is implemented and (in my testing) functional in 4.2.1-stable.

I'll also confirmed that arm64 binaries can be built for the `3.x` branch, so we should be able to have this in 3.6 too.

**Comment by Schweini07** [comment]
This is very cool to see, after all this time :smile:

**Comment by Calinou** [comment]
> Whenever I execute it through the console or through the file, it works fine until I try to open a project. When I do, it says Godot game engine for a second then crashes. What am I doing wrong?

Vulkan support on the Raspberry Pi is still spotty, so you need to create/open projects using the Compatibility rendering method. If the project is configured to use another rendering method, open the project directly on the [command line](https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html) using the `--rendering-method gl_compatibility` command line argument.

**Comment by RiptideDev** [comment]
@Calinou Works perfect now, thanks so much!

**Comment by real-felix** [comment]
@akien-mga It works really well! (I'm testing it on Asahi Linux). Do you know if it'll be available on the official page, with the latest version, in the future?

**Comment by Calinou** [comment]
> @akien-mga It works really well! (I'm testing it on Asahi Linux). Do you know if it'll be available on the official page, with the latest version, in the future?

This was added today in https://github.com/godotengine/godot-website/pull/899.

**Comment by sosasees** [comment]
will i be able to export to
- Linux/X11 (arm)
- Linux/X11 (arm64)

from Linux/X11 (x86_64) in an upcoming Godot version?

**Comment by Calinou** [comment]
> from Linux/X11 (x86_64) in an upcoming Godot version?

It's already feasible by checking the correct architecture options, as long as you have export templates installed.
